### ISSUE TITLE
Feature Request: Allow Arrays To Be Expanded By Default

### ISSUE THREAD

**Issue by Undistraction** [prettier/prettier#817]
Please add a configuration option to allow arrays to be formatted as expanded, regardless of the number of items or the length of the line.

Yay
```
const example = [
  alpha,
  beta,
  charlie,
  delta,
  echo,
];
```
Nay
```
const example = [alpha, beta, charlie, delta, echo];
```

There are a number of reasons why this is beneficial:

1. **Readability**: an array of items is more readable when it can be scanned as a list. Yes we read left to right, but when it comes to lists we read top to bottom. Bullet points are the perfect example of this. This is particularly true of sequences where order is important.

2. **Consistency**: having all arrays behave the same regardless of content is much more consistent.

3. **Ease of use**: it is much easier to move up and down a vertical list with a keyboard. It is also easier to duplicated a line. 

3. **Reviewability**: It is much easier to see what has changed in a diff if there is one item on every line.

These same arguments can also be applied to restructuring of imports and variables.

Yay
```
{
  alpha,
  beta,
  charlie,
  delta,
  echo,
} = example;
```
Nay
```
{ alpha, beta, charlie, delta, echo } = example;
 ```

**Comment by Undistraction** [comment]
Thinking about this further, the same arguments apply to React component attributes:

```
<Toggle
      className={styles.autoToggle}
      id="autoToggle"
      label="Auto"
      onChange={isSelected => onAutoChange(isSelected)}
      isSelected={isAuto}
    />
```

@jlongster Would you prefer this as a separate issue? As the arguments are the same I think it makes sense to include in this issue?

**Comment by jlongster** [comment]
The way we do this is not to force it to be expanded, but to make it expand *if* the original object was already expanded. So it'll stay collapsed if it's like that in the original source.

We really should not be expanding things most of the time though. It's super annoying because it'll expand something once it gets large enough, but if you remove something immediately afterwards it doesn't collapse it again. The way this projects works is to collapse things when it can, and we're not going to make a lot of exceptions.

We're definitely not going to do it for JSX attributes, but we can talk about arrays. I'm not sure we should add that exception either but we can leave this issue open to see what other think.

**Comment by Undistraction** [comment]
@jlongster So what are the benefits to always collapsing arrays? I think the four points I've outlined above are a very strong argument against. The only counter argument I can see is for readability, but I think even that is very dubious. Am I missing some obvious benefit(s)? 

Apart from anything else I would argue treatment of Objects and Arrays should be consistent.

**Comment by jlongster** [comment]
Your points are more subjective than you think. Also, you run into collapsed arrays all the time for simple stuff like `foo(["x", "y", "z"])`. We certainly don't want to expand that as I think it aligns more with all your points than not. If that is deep in an expression, expanding it will make the entire larger expression break which is very undesirable.

It seems like you generally want anything that is list-like to always be expanded, and we are not going to do that because that affects the style far too much. It does make it harder to read and review in some cases. But like I said, nothing here is objective, it's all subjective, and we have to make a call and I think the way it currently works is what we're going to stick to.

**Comment by Undistraction** [comment]
Thanks for taking the time to respond. 

Just to be clear - I wasn't suggesting this should be forced on anyone. It's just that I think there is enough of an argument (and enough people doing this in the wild) that is warrants a configuration option. Obviously you have a clear idea what the trade-off is as far as additional work and complexity, but I hope you will decide it is worth it.

**Comment by rhengles** [comment]
@Undistraction Alright, this is a little weird even for me. But I created an option for it on https://github.com/arijs/prettier-with-tabs/commit/f6c12c62a6afcb64f997f2ea8d064df440f1b4fd - please check out if the result is what you expected 😉 

@jlongster @vjeux and prettier team - when you get esoteric feature requests for unconventional options, please feel free to forward them to me 😃

**Comment by jlongster** [comment]
@Undistraction we're really trying to limit configuration options, and the ones we have are very small tweaks that affect tiny things like single/double quotes and trailing commas. We're not going to have a config option that drastically affects the output like expanding certain types, unfortunately.

It's good that @rhengles is around to provide a lot more options :) Thanks for taking on more esoteric requests!

Thank you for understanding and I'm sorry we can't fit your needs. :(

**Comment by jlongster** [comment]
@vjeux What do you think about arrays though? The more I think about it, I don't work with array literals too often so if we want to be consistent with objects, we could do that. It would probably make some people happy when dealing with config files.

**Comment by vjeux** [comment]
I don't really care either way. Object was the thing people complained on a daily basis and was a deal breaker. Array is much less of an issue in practice.

**Comment by marianoguerra** [comment]
hi, I installed prettier today (0.22.0) and tried it on a requirejs project and it expanded the list of deps on the top level define to be one per line, but then the arg list of the function was on one line, that seems inconsistent and generates really tall lists for "central" modules, here's an example of the input:

    define(['aaaaaaaaaaaaaaaa', 'bbbbbbbbbbbbbbbbbb', 'ccccccccccccccccc', 'ddddddddddddddddd'], function(aaaaaaaaaaaaaaaa, bbbbbbbbbbbbbbbbbb, ccccccccccccccccc, ddddddddddddddddd) {});

result:

        define(
          [
            'aaaaaaaaaaaaaaaa',
            'bbbbbbbbbbbbbbbbbb',
            'ccccccccccccccccc',
            'ddddddddddddddddd',
          ],
          function(aaaaaaaaaaaaaaaa, bbbbbbbbbbbbbbbbbb, ccccccccccccccccc, ddddddddddddddddd) {}
        );

I think having as many items per line is better and consistent with function arguments.

**Comment by vjeux** [comment]
We shipped prettier 1.0 and are not going to do major formatting changes anymore. Sorry :( If you really want this feature, I would recommend you to fork prettier.
